compiler/rustc_mir/src/transform/basic_block_copy_prop.rs

```diff
@@ -0,0 +1,361 @@
+        self.super_statement(statement, location);
+
+        // Now that we've invalidated locals mutated by the statement, replace any known locals that
+        // *aren't* mutated with the place they were assigned to.
+        ReplaceUseVisitor { tcx: self.tcx, local_values: &self.local_values }
+            .visit_statement(statement, location);
+
+        InvalidateMovedLocals { tcx: self.tcx, local_values: &mut self.local_values }
+            .visit_statement(statement, location);
+    }
+
+    fn visit_terminator(&mut self, terminator: &mut Terminator<'tcx>, location: Location) {
+        self.super_terminator(terminator, location);
+
+        // Now that we've invalidated locals mutated by the terminator, replace any known locals
+        // that *aren't* mutated with the place they were assigned to.
+        ReplaceUseVisitor { tcx: self.tcx, local_values: &mut self.local_values }
+            .visit_terminator(terminator, location);
+
+        InvalidateMovedLocals { tcx: self.tcx, local_values: &mut self.local_values }
+            .visit_terminator(terminator, location);
+    }
+
+    fn visit_local(&mut self, local: &mut Local, context: PlaceContext, location: Location) {
+        // We invalidate a local `l`, and any other locals whose assigned values contain `l`, if:
+        // - `l` is mutated (via assignment or other stores, by taking a mutable ref/ptr, etc.)
+        // - `l` is reallocated by storage statements (which deinitializes its storage)
+
+        if context.is_mutating_use() || context.is_storage_marker() {
+            debug!("invalidation of {:?} at {:?}: {:?} -> clearing", local, location, context);
+            self.local_values.invalidate(*local);
+        }
+
+        // Note that we do not yet do anything when `l` is moved out of. The reason for that is that
+        // it is fine to replace that local (see reasoning below). We just have to make sure we
+        // invalidate the moved-out-of local *after* we finished processing the statement/terminator
+        // performing the move.
+    }
+}
+
+struct ReplaceUseVisitor<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    local_values: &'a LocalValues<'tcx>,
+}
+
+impl<'a, 'tcx> MutVisitor<'tcx> for ReplaceUseVisitor<'a, 'tcx> {
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+
+    fn visit_operand(&mut self, operand: &mut Operand<'tcx>, location: Location) {
+        // NB: All `operand`s are non-mutating uses of the contained place, so we don't have to call
+        // `super_operand` here.
+
+        // We *can* replace a `move` by a copy from the recorded place, because we only record
+        // places that are *copied* from in the first place (so the place type must be `Copy` by
+        // virtue of the input MIR). For example, this is a common pattern:
+        //   _1 = ...;
+        //   _2 = _1;
+        //   use(move _2);   <- can replace with `use(_1)`
+        // Or:
+        //   use(move _2, move _1); <- can replace with `use(_1, _1)`
```

Comment:
Hmm, this doesn't sound right, this is only correct if the places `_1` and `_1` are used are allowed to overlap

Reply:
Are there cases where they are not?

From codegen perspective it seemed correct to me when I was reviewing the changes. Based on the fact that codegen for `Operand::Copy` introduces a copy when passing by ref, unlike `Operand::Move` which does not.

Reply:
Yeah, for example call destinations must not alias any call arguments (maybe this only applies to copied arguments, as you mentioned). I ran into this when writing the destination propagation pass.

Unfortunately these invariants haven't really been consciously designed or documented from what I know, but the MIR validator does try to enforce them (or what I think is a reasonably approximation of them).

It seems that this pass is still correct though – aliasing restrictions generally only apply when one of the places is being mutated, and when that happens we invalidate all affected locals.

Reply:
Extended the comment

Reply:
There is also implicit assumption about when the move invalidates the local. As far as I can see this pass could make a following transformation:

```
_2 = _1
_3 = f(move _1, move _2);
```

```
_2 = _1
_3 = f(move _1, _1);
```

Reply:
Oof, yeah, that doesn't sound like it *should* work, but maybe it currently does. Promotion of copies to moves is something we do want to do at some point, so that would have to agree with this pass.